runtime/codegen/encoder.go

```diff
@@ -250,23 +250,86 @@ func (e *Encoder) Len(l int) {
 	e.Int32(int32(l))
 }
 
+//	Error encoding
+//
+// An error can be composed of a tree of errors (see the errors package)
+// We encode such a tree in a linear list of errors visited in depth-first
+// order.
+//
+// The encoding is a list of encoded errors terminated by a <endOfErrors>.
+// An encoded error is either <serializedError,typKey,serial> for registered
+// serializable error types, or <emulatedError,message,fmtError> for
+// unregistered error types.
+const (
+	endOfErrors     uint8 = 0
+	serializedError uint8 = 1
+	emulatedError   uint8 = 2
+)
+
 // Error encodes an arg of type error. We save enough type information
-// to allow errors.Unwrap() and errors.Is() to work correctly.
+// to allow errors.Unwrap(), errors.Is(), and errors.As() to work correctly.
 func (e *Encoder) Error(err error) {
-	// Get the stack of wrapped errors.
-	stack := make([]error, 0, 4)
-	for err != nil {
-		stack = append(stack, err)
-		err = errors.Unwrap(err)
-	}
-
-	e.Int(len(stack))
-	for _, err := range stack {
+	// Convert the tree of wrapped errors into a single list.
+	// We do not use errors.Unwrap() since it does not visit the
+	// children found by "Unwrap()[]error".
+	//
+	// The encoding is a list of encoded errors terminated by a <0>.
+	// An encoded error is either <1,typKey,serialization> for
+	// registered custom error types, or <2,message,fmtError> for
+	// unregistered error types.
+	var dfs func(error)
+	dfs = func(err error) {
```

Comment:
Do you think it's eventually worth worrying about cyclic errors (e.g., https://go.dev/play/p/_6g1xLGBQ6f) that could make this code infinite loop? I think we currently disallow recursive types, but error "trees" can be cyclic. I guess even calling the `Error` method on an error could infinite loop, so maybe it's a moot point.